sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -238,6 +249,11 @@ class BulkWriter(container: CosmosAsyncContainer,
 
         log.logInfo(s"completed so far ${totalSuccessfulIngestionMetrics.get()}, pending tasks ${activeOperations.size}")
 
+        // Closing the input flux will cause the BulkExecutor to flush all buffered
+        // item operations and start a timer that would flush regularly
+         log.logInfo("invoking bulkInputEmitter.onComplete()")
+        bulkInputEmitter.tryEmitComplete()
```

Comment:
I wonder if there is a specific reason why this is moved before the lock section?

Reply:
as discussed offline, I think we might be having a deadlock due to retires